File: cpp/insertionsort.cpp
#include <iostream>
#include <vector>
using namespace std;

void printVector(const vector<int>& arr) {
    for (int num : arr) {
        cout << num << " ";
    }
    cout << endl;
}

void insertionSort(vector<int>& arr) {
    int size = arr.size();
    for (int i = 1; i < size; i++) {
        int key = arr[i];
        int j = i - 1;

        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;

        cout << "After pass " << i << ": ";
        printVector(arr);
    }
}

int main() {
    int size;
    cout << "Enter the size of array: ";
    cin >> size;

    vector<int> arr(size);
    cout << "Enter " << size << " elements: ";
    for (int i = 0; i < size; i++) {
        cin >> arr[i];
    }

    cout << "Original array: ";
    printVector(arr);

    cout << "\nSorting process:\n";
    insertionSort(arr);

    cout << "\nSorted array: ";
    printVector(arr);

    return 0;
}

File: python/insertionsort.py
def print_array(arr, size):
    for i in range(size):
        print(arr[i], end=" ")
    print()

def insertion_sort(arr, size):
    for i in range(1, size):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j = j - 1
        arr[j + 1] = key
        print_array(arr, size)


size = int(input("Enter the size of array: "))
arr = [0] * size 
    
print(f"Enter {size} elements: ", end="")
elements = input().split()
for i in range(size):
    arr[i] = int(elements[i])
    
print("Original array: ", end="")
print_array(arr, size)
    
insertion_sort(arr, size)
    
print("Sorted array: ", end="")
print_array(arr, size)
